File: src/main/java/com/farthergate/ctrlcurve/PID.java
package com.farthergate.ctrlcurve;
import com.farthergate.ctrlcurve.math.Tolerance;

public class PID {
    double kp, ti, td, dt
         , t = 0
         , error
         , tolerance
         , initial
         , target
         , current
         , previousError = 0
         , integral = 0;

    public PID(double kp, double ti, double td, double dt, double tolerance, double initial, double target) {
        this.kp = kp;
        this.ti = ti;
        this.td = td;
        this.dt = dt;
        this.tolerance = tolerance;
        this.initial = initial;
        this.target = target;
        current = initial;
    }
    
    public boolean shouldContinue() {
        return !Tolerance.isWithinTolerance(target, current, tolerance);
    }

    public double correction() {
        var error = target - current;
        var proportional = error;
        integral = integral + error*dt;
        var derivative = (error - previousError) / dt;
        previousError = error;
        return kp * (proportional + integral/ti + td*derivative/1000);
    }

    public void sync() throws InterruptedException {
        Thread.sleep((long)dt);
        t += dt;
    }
}

File: lib/bin/main/com/rutins/aleks/ctrlcurve/pid.kt
package com.rutins.aleks.ctrlcurve

typealias PIDCorrection = PIDScope.(initial: Double, current: Double, target: Double, correction: Double) -> Double

class PIDScope(val k: Double, var error: Double, val tolerance: Double, val initial: Double, val target: Double) {
    var stopRequested = false
    fun calculateCorrection(): Double {
        if (k != 0.0 && (target-initial) != 0.0) {
            return k * (error / (target - initial))
        }
        return target
    }
}

fun runPID(initial: Double, target: Double, tolerance: Double, k: Double, correction: PIDCorrection): Double {
    var current = initial
    var scope = PIDScope(k, target - current, tolerance, initial, target)
    while(!isWithinTolerance(current, target, tolerance) && !scope.stopRequested) {
        scope.error = target - current
        current = scope.correction(initial, current, target, target - current)
    }
    return current
}